### PR TITLE
packaging: move from setup.py to setup.cfg

### PR DESCRIPTION
**Pull request by lnielsen**
### Description

Implements [RFC #57](https://github.com/inveniosoftware/rfcs/blob/1b6ccb599a17b97bca02a7c95e0eed340dc42c70/rfcs/framework-0057-declarative-python-package-definition.md)

### Checklist
- [x] Did you create different commits for logically separate things? Check [the format](https://invenio.readthedocs.io/en/latest/community/contributing/contribution-guide.html#commit-messages) of the commit(s) message(s).
- [x] [InvenioRDM specific] Did you follow the style guide?  Check [the documentation](https://inveniordm.docs.cern.ch/maintenance/styleguide/).
- [x] [InvenioRDM specific] Did you add translations (i18n) for new labels, backend and frontend? Check [the documentation](https://inveniordm.docs.cern.ch/develop/topics/i18n/).
- [x] [InvenioRDM specific] Did you add accessibility (a11y) tags? Check [the documentation](https://inveniordm.docs.cern.ch/maintenance/accessibility/).
- [x] Did you add documentation? We all love well-documented frameworks, right?
- [ ] Did you include third-party code (a library, package, etc.)? If yes:
    - [ ] Make sure that its open source license is compatible with Invenio (*note:* GPL and AGPL licensed code is not accepted)
    - [ ] Did you respect the license conditions (normally you need to keep the copyright and license header)?

> Green light on all GitHub Actions status checks is required in order to merge your PR. Check your code additions will fail neither code linting checks nor tests and will not decrease test coverage.

#### Copyright

- [x] Identify who is the copyright holder (a legal entity - i.e. person or organization - normally this is your employer) for the produced code.
- [x] Make sure that you have updated the copyright headers for touched files (non-trivial contributions >15 lines).

See [copyright and license policy](https://invenio.readthedocs.io/en/latest/community/contributing/maintainers-guide/copyright-policy.html) for details.

#### Reminder

By using GitHub, you have already agreed to the [GitHub’s Terms of Service](https://help.github.com/articles/github-terms-of-service/#6-contributions-under-repository-license) including that:

1. You license your contribution under the same terms as the current repository’s license.
2. You agree that you have the right to license your contribution under the current repository’s license.

If you ticked all the boxes above, you're good to go! :rocket:

### CODE REVIEW

**Comment by lnielsen** [comment]
asdfsdfsadf

File: {{ cookiecutter.project_shortname }}/.github/workflows/tests.yml [hunk 1]
```diff
@@ -1,95 +1,50 @@
...
-          # a new item to the following lists.
+          # TODO: You can add/remove combinations e.g. `dev` requirements or
+          # `postgresql13` by adding a new item to the following lists.
           # You can see the complete list of services and versions that are available at:
           # https://docker-services-cli.readthedocs.io/en/latest/configuration.html
-          python-version: [3.7, 3.8, 3.9, "3.10"]
-          requirements-level: [min, pypi]
-          cache-service: [redis]
-          db-service: [postgresql10, postgresql13, mysql5, mysql8]
-          mq-service: [rabbitmq, redis]
```
Comment (slint) [inline {{ cookiecutter.project_shortname }}/.github/workflows/tests.yml:32]: - Removed `cache-service` since we only support redis
- Removed `mq-service` since we only support RabbitMQ

File: {{ cookiecutter.project_shortname }}/.github/workflows/tests.yml [hunk 2]
```diff
@@ -1,95 +1,50 @@
...
-
-          - db-service: postgresql13
-            DB_EXTRAS: "postgresql"
-
-          - db-service: mysql5
-            DB_EXTRAS: "mysql"
-
-          - db-service: mysql8
-            DB_EXTRAS: "mysql"
+          - search-service: opensearch2
```
Comment (slint) [inline {{ cookiecutter.project_shortname }}/.github/workflows/tests.yml:37]: Only kept search-related extras, since for db we install all database drivers anyways in `invenio-db`

File: {{ cookiecutter.project_shortname }}/setup.cfg [hunk 3]
```diff
@@ -17,15 +17,22 @@ classifiers =
...
 install_requires =
     invenio-base>=1.2.5
     invenio-i18n>=1.3.1
 
 [options.extras_require]
 tests =
-    pytest-invenio>=1.4.2
-    sphinx>=4.2.0
+    pytest-invenio>=2.1.0,<3.0.0
+    pytest-black>=0.3.0
```
Comment (slint) [inline {{ cookiecutter.project_shortname }}/setup.cfg:29]: I opened https://github.com/inveniosoftware/pytest-invenio/issues/92 to see if we could integrate in `pytest-invenio` and skip it.

File: {{ cookiecutter.project_shortname }}/setup.cfg [hunk 4]
```diff
@@ -59,6 +66,17 @@ add_ignore = D401
...
 
+[isort]
+profile=black
+
+[check-manifest]
+ignore =
+    *-requirements.txt
+
+[tool:pytest]
+addopts = --black --isort --pydocstyle --pycodestyle --doctest-glob="*.rst" --doctest-modules --cov={{ cookiecutter.package_name }} --cov-report=term-missing
```
Comment (slint) [inline {{ cookiecutter.project_shortname }}/setup.cfg:77]: I saw mixed results in repos regarding `pycodestyle`. We should probably review linting and what kind of tools we want to use/enforce

Comment (utnapischtim) [reply {{ cookiecutter.project_shortname }}/setup.cfg:77]: for black managed packages the `--pycodestyle` has been removed. If i remember correctly there were problems with black and pycodestyle. Further in my opinion pycodestyle is useless if we have an opinionated code formatter

Comment (slint) [reply {{ cookiecutter.project_shortname }}/setup.cfg:77]: Thanks for chipping-in, I couldn't figure out if that was a pattern or random :)

In that case, I'm removing pycodestyle from the pytest flags and any other config.

I was also wondering if we could replace `flake8`, `pycodestyle`, etc. with [`ruff`](https://github.com/charliermarsh/ruff), which is the new kid in the Python linters block:

- written in Rust for performance (10-100x compared to the above tools)
- compatible with `black` [out of the box](https://github.com/charliermarsh/ruff#is-ruff-compatible-with-black)
- widely adopted but big projects

Something for much later though :)

File: {{ cookiecutter.project_shortname }}/setup.cfg [hunk 5]
```diff
@@ -11,8 +63,16 @@ universal = 1
 [pydocstyle]
 add_ignore = D401
 
-[pycodestyle]
-exclude = docs/conf.py
+[isort]
+profile=black
+
+[check-manifest]
```
Comment (utnapischtim) [inline {{ cookiecutter.project_shortname }}/setup.cfg:69]: it should be changed the check_manifest line in run-tests.sh too
from `python -m check_manifest --ignore ".*-requirements.txt"` to `python -m check_manifest`

File: {{ cookiecutter.project_shortname }}/setup.cfg (hunk 4 above)
Comment (utnapischtim) [reply {{ cookiecutter.project_shortname }}/setup.cfg]: i commented on the [inveniosoftware/pytest-invenio#93](https://github.com/inveniosoftware/pytest-invenio/issues/93) about the idea to use `ruff` instead of `pycodestyle` and `flake8`

File: {{ cookiecutter.project_shortname }}/setup.cfg [hunk 6]
```diff
@@ -1,4 +1,57 @@
...
+zip_safe = False
+install_requires =
+    invenio-base>=1.2.5
+    invenio-i18n>=1.3.1
+
+[options.extras_require]
+tests =
+    invenio-app>=1.3.4,<2.0.0
+    pytest-invenio>=2.1.0,<3.0.0
+    pytest-black>=0.3.0
```
Comment (ppanero) [inline {{ cookiecutter.project_shortname }}/setup.cfg:30]: minor/shelvable: should this be moved to pytest-invenio since now all modules support black?

File: {{ cookiecutter.project_shortname }}/setup.cfg [hunk 7]
```diff
@@ -1,4 +1,57 @@
...
+packages = find:
+python_requires = >=3.7
+zip_safe = False
+install_requires =
+    invenio-base>=1.2.5
+    invenio-i18n>=1.3.1
+
+[options.extras_require]
+tests =
+    invenio-app>=1.3.4,<2.0.0
```
Comment (ppanero) [inline {{ cookiecutter.project_shortname }}/setup.cfg:28]: moderate: I thought we were only putting an upper cap in `invenio-app-rdm` and in cases where there is breaking changes releases by 3rd parties. Otherwise new modules will come with all invenio dependencies capped.

**Review by ppanero — APPROVED** [review]
Assuming the comments are addressed/answered feel free to merge!

File: {{ cookiecutter.project_shortname }}/setup.cfg (hunk 6 above)
Comment (slint) [reply {{ cookiecutter.project_shortname }}/setup.cfg:30]: Same thought: https://github.com/inveniosoftware/cookiecutter-invenio-module/pull/147#discussion_r1101239515

I can open an issue in `pytest-invenio` and we can address it there.

Comment (ppanero) [reply {{ cookiecutter.project_shortname }}/setup.cfg:30]: Yes! Sorry, it was closed/resolved and I missed it

File: {{ cookiecutter.project_shortname }}/setup.cfg (hunk 7 above)
Comment (slint) [reply {{ cookiecutter.project_shortname }}/setup.cfg:28]: We're notoriously inconsistent regarding the version capping on modules, and I can understand that since there are good pros/cons for either choice. I think the major ones are:

+ Better stability: capping prevents modules from breaking when we release breaking changes (i.e. major bump) on the dependency
- More work: without the tests breaking when a dependency has breaking changes, we'll never find out the impact of a breaking change unless we manually check each module that depends on it

I've found a mix of capped/uncapped across base Invenio and RDM modules. Maybe a general "pattern" is that the higher up in the chain a module is or the more critical its dependencies are, the more likely they're capped:

- [rdm-records](https://github.com/inveniosoftware/invenio-rdm-records/blob/08c420f4745a9979cb864ef33d38fdb7f5096138/setup.cfg#L39-L43)
- [records-resources](https://github.com/inveniosoftware/invenio-records-resources/blob/389a96f8fe22103e7ee5ad3ff19c682402403e46/setup.cfg#L30-L39)

I would say given past situations, capping is a more stable alternative that helps also with overall developer experience.

Comment (ppanero) [reply {{ cookiecutter.project_shortname }}/setup.cfg:28]: I agree with you in pros/cons. However, I tend to lean more on controlled chaos and cap when we notice breaking changes (true that we should have better mechanisms to do so). I'm okay with merging this as-is. However, since this will affect ALL new modules maybe we should raise the topic to a larger audience (thursday? architects)?

TL;DR: feel free to merge, maybe worth discussing with broader audience.
